lib/Basic/Fingerprint.cpp

```diff
@@ -25,19 +29,26 @@ void swift::simple_display(llvm::raw_ostream &out, const Fingerprint &fp) {
   out << fp.getRawValue();
 }
 
-namespace {
-  template <class T> struct SmallStringBound;
-  template <size_t N> struct SmallStringBound<llvm::SmallString<N>> {
-    static constexpr size_t value = N;
-  };
-};
+Fingerprint Fingerprint::fromString(StringRef value) {
+  assert(value.size() == Fingerprint::DIGEST_LENGTH &&
+         "Only supports 32-byte hash values!");
+  auto fp = Fingerprint::ZERO();
+  {
+    std::istringstream s(value.drop_back(Fingerprint::DIGEST_LENGTH/2).str());
```

Comment:
What happens if DIGEST_LENGTH is odd? Should there be an assert-even here?

Reply:
It cannot be odd. It's a constant - it's always going to be 32 bytes. See the header.

Reply:
Yes, but the point is to centralize the invariant, in case someday, someone changes the header, and then someone else changes what it depends upon, etc.